module/zfs/dmu.c

```diff
@@ -2273,19 +2280,23 @@ dmu_brt_clone(objset_t *os, uint64_t object, uint64_t offset, uint64_t length,
 		ASSERT(db->db_blkid != DMU_BONUS_BLKID);
 		ASSERT(BP_IS_HOLE(bp) || dbuf->db_size == BP_GET_LSIZE(bp));
 
-		if (db->db_state == DB_UNCACHED) {
-			/*
-			 * XXX-PJD: If the dbuf is already cached, calling
-			 * dmu_buf_will_not_fill() will panic on assertion
-			 * (db->db_buf == NULL) in dbuf_clear_data(),
-			 * which is called from dbuf_noread() in DB_NOFILL
-			 * case. I'm not 100% sure this is the right thing
-			 * to do, but it seems to work.
-			 */
-			dmu_buf_will_not_fill(dbuf, tx);
+		mutex_enter(&db->db_mtx);
+
+		VERIFY(!dbuf_undirty(db, tx));
+		ASSERT(list_head(&db->db_dirty_records) == NULL);
+		if (db->db_buf != NULL) {
+			arc_buf_destroy(db->db_buf, db);
+			db->db_buf = NULL;
 		}
 
+		mutex_exit(&db->db_mtx);
+
+		dmu_buf_will_not_fill(dbuf, tx);
+
+		mutex_enter(&db->db_mtx);
+
 		dr = list_head(&db->db_dirty_records);
+		VERIFY(dr != NULL);
```

Comment:
I wonder if a severe scheduling delay could cause us to see a NULL pointer here. I recall 4c856fb333ac57d9b4a6ddd44407fd022a702f00 where a race existed that was only ever tripped on Windows because the scheduler there less fair than UNIX schedulers.

Reply:
This is easily fixable by just retrying in this situation. I've updated the pull request.

Reply:
Isn't the potential problem that the dbuf can be dirtied again once we drop the db_mtx lock? We already deal with situation where we clone a block and in the same txg we overwrite cloned block with real data.